### ISSUE TITLE
azurerm_monitor_data_collection_rule not possible to create Custom Log

### ISSUE THREAD

**Issue by dansmitt** [hashicorp/terraform-provider-azurerm#21897]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

latest

### AzureRM Provider Version

3.57.0

### Affected Resource(s)/Data Source(s)

azurerm_monitor_data_collection_rule

### Terraform Configuration Files

```hcl
resource "azurerm_resource_group" "example" {
  name     = "example-resources"
  location = "West Europe"
}

resource "azurerm_log_analytics_workspace" "auditlogla" {
  name                = "auditlogla"
  location            = azurerm_resource_group.location
  resource_group_name = azurerm_resource_group.name
  sku                 = "PerGB2018"
  retention_in_days   = 30
}

resource "azurerm_monitor_data_collection_endpoint" "dce" {
  name                = "example-dce"
  location            = azurerm_resource_group.location
  resource_group_name = azurerm_resource_group.name

  lifecycle {
    create_before_destroy = true
  }
}

resource "azurerm_monitor_data_collection_rule" "dcr" {
  name                        = "example-dcr"
  location                    = azurerm_resource_group.location
  resource_group_name         = azurerm_resource_group.name
  data_collection_endpoint_id = azurerm_monitor_data_collection_endpoint.dce.id

  destinations {
    log_analytics {
      workspace_resource_id = azurerm_log_analytics_workspace.auditlogla.id
      name                  = "example-destination-log"
    }
  }

  data_flow {
    streams       = ["Custom-MyTableRawData"]
    destinations  = ["example-destination-log"]
    output_stream = "Custom-MyTable_CL"
    transform_kql = "source | project TimeGenerated = Time, Computer, Message = AdditionalContext"
  }

  stream_declaration {
    stream_name = "Custom-MyTableRawData"
    column {
      name = "Time"
      type = "datetime"
    }
    column {
      name = "Computer"
      type = "string"
    }
    column {
      name = "AdditionalContext"
      type = "string"
    }
  }

  depends_on = [
    azurerm_log_analytics_workspace.auditlogla
  ]
}
```

### Debug Output/Panic Output

```shell
Service returned an error. Status=400 Code="InvalidPayload" Message="Data collection rule is invalid" Details=[{"code":"InvalidOutputTable","message":"Table for output stream 'Custom-MyTable_CL' is not available for destination 'example-destination-log'.","target":"properties.dataFlows[0]"}]
```

### Expected Behaviour

Custom-MyTable_CL should be created before the stream will be created.

### References

#21880

**Comment by simaotwx** [comment]
Why is stream_declaration necessary in the resource when it needs to be created first anyway?

**Comment by teowa** [comment]
From my obeservation, if we need to use a custom stream in the `streams` field, we need to specify it in `stream_declaration` first.

**Comment by simaotwx** [comment]
@teowa I tried to use your example as reference for integrating it, but I'm having a lot of trouble with Azure API (which is often a pain to work with):
```
│ --------------------------------------------------------------------------------
│ RESPONSE 500: 500 Internal Server Error
│ ERROR CODE UNAVAILABLE
│ --------------------------------------------------------------------------------
│ Response contained no body
│ --------------------------------------------------------------------------------
│ 
│ 
│   with azapi_resource.data_collection_raw_monolog_logs_table,
│   on log_analytics.tf line 26, in resource "azapi_resource" "data_collection_raw_monolog_logs_table":
│   26: resource "azapi_resource" "data_collection_raw_monolog_logs_table" {
│ 
╵
```

My resource:
```
resource "azapi_resource" "data_collection_raw_monolog_logs_table" {
  name      = "RawMonologLogs_CL"
  parent_id = azurerm_log_analytics_workspace.log_analytics_workspace.id
  type      = "Microsoft.OperationalInsights/workspaces/tables@2022-10-01"
  body = jsonencode(
    {
      "properties" : {
        "schema" : {
          "name" : "RawMonologLogs_CL",
          "columns" : [
            {
              "name" : "LogTime",
              "type" : "datetime",
              "description" : "The time at which the data was generated"
            },
            {
              "name" : "Level",
              "type" : "string",
              "description" : "Log level"
            },
            {
              "name" : "Logger",
              "type" : "string",
              "description" : "Logger name"
            },
            {
              "name" : "Context",
              "type" : "string",
              "description" : "Context of the log line"
            },
            {
              "name" : "Message",
              "type" : "string",
              "description" : "Log message"
            },
            {
              "name" : "AdditionalContext",
              "type" : "string",
              "description" : "Additional context of the log line"
            }
          ]
        }
      }
    }
  )
}
```

What's really weird (and causes a lot of frustration) is that the resource is **actually created** but Azure still returns Internal Server Error (not the first time something like this happens) which causes inconsistency in the Terraform state, because next time I try to run the apply, this is the response:
```
azapi_resource.data_collection_raw_monolog_logs_table: Creating...
╷
│ Error: A resource with the ID "/subscriptions/***/resourceGroups/***/providers/Microsoft.OperationalInsights/workspaces/***/tables/RawMonologLogs_CL" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azapi_resource" for more information.
│ 
│   with azapi_resource.data_collection_raw_monolog_logs_table,
│   on log_analytics.tf line 26, in resource "azapi_resource" "data_collection_raw_monolog_logs_table":
│   26: resource "azapi_resource" "data_collection_raw_monolog_logs_table" {
│ 
╵
```

I tried importing the resource, which worked. But then, upon apply, this happens:

```
azapi_resource.data_collection_raw_monolog_logs_table: Modifying... [id=...]
azapi_resource.data_collection_raw_monolog_logs_table: Still modifying... [id=... 10s elapsed]
azapi_resource.data_collection_raw_monolog_logs_table: Still modifying... [id=..., 20s elapsed]
azapi_resource.data_collection_raw_monolog_logs_table: Still modifying... [id=..., 30s elapsed]
azapi_resource.data_collection_raw_monolog_logs_table: Still modifying... [id=..., 40s elapsed]
╷
│ Error: creating/updating "Resource: (ResourceId \"/subscriptions/***/resourceGroups/***/providers/Microsoft.OperationalInsights/workspaces/***/tables/RawMonologLogs_CL\" / Api Version \"2022-10-01\")": PUT https://management.azure.com/subscriptions/***/resourceGroups/***/providers/Microsoft.OperationalInsights/workspaces/***/tables/RawMonologLogs_CL
│ --------------------------------------------------------------------------------
│ RESPONSE 500: 500 Internal Server Error
│ ERROR CODE UNAVAILABLE
│ --------------------------------------------------------------------------------
│ Response contained no body
│ --------------------------------------------------------------------------------
│ 
│ 
│   with azapi_resource.data_collection_raw_monolog_logs_table,
│   on log_analytics.tf line 26, in resource "azapi_resource" "data_collection_raw_monolog_logs_table":
│   26: resource "azapi_resource" "data_collection_raw_monolog_logs_table" {
│ 
╵
```

The plan said following:
```

  # azapi_resource.data_collection_raw_monolog_logs_table will be updated in-place
  ~ resource "azapi_resource" "data_collection_raw_monolog_logs_table" {
      ~ body                      = jsonencode(
          ~ {
              ~ properties = {
                  - plan                 = "Analytics" -> null
                  - retentionInDays      = 30 -> null
                  ~ schema               = {
                      ~ columns = [
                          + {
                              + description = "The time at which the data was generated"
                              + name        = "LogTime"
                              + type        = "datetime"
                            },
                          + {
                              + description = "Log level"
                              + name        = "Level"
                              + type        = "string"
                            },
                          + {
                              + description = "Logger name"
                              + name        = "Logger"
                              + type        = "string"
                            },
                          + {
                              + description = "Context of the log line"
                              + name        = "Context"
                              + type        = "string"
                            },
                          + {
                              + description = "Log message"
                              + name        = "Message"
                              + type        = "string"
                            },
                          + {
                              + description = "Additional context of the log line"
                              + name        = "AdditionalContext"
                              + type        = "string"
                            },
                        ]
                        name    = "RawMonologLogs_CL"
                    }
                  - totalRetentionInDays = 30 -> null
                }
            }
        )
        id                        = "/subscriptions/***/resourceGroups/***/providers/Microsoft.OperationalInsights/workspaces/***/tables/RawMonologLogs_CL"
        name                      = "RawMonologLogs_CL"
      ~ output                    = jsonencode({}) -> (known after apply)
        tags                      = {}
        # (6 unchanged attributes hidden)

      - timeouts {}
    }
```

**Comment by simaotwx** [comment]
I looked in Azure Portal and the table is not listed there, so this is starting to get really weird.

**Comment by teowa** [comment]
Hi @simaotwx , this might because [Azapi currently not wait for long-running-operation](https://github.com/Azure/terraform-provider-azapi/issues/281), and the create table operation is a [long-running-operation](https://github.com/Azure/azure-rest-api-specs/blob/cd06d327e115cdb55f8e7c9fd1b23fa551b5b750/specification/operationalinsights/resource-manager/Microsoft.OperationalInsights/stable/2022-10-01/Tables.json#L137), this means when we send request to create the table, the creation take some time to finish after the API respose return. And there might be portal display lantency for the newly created table. For this we can wait some time for the creation to finish. 
And for the update failure (500 error code), it should be service API bugs, to workaround this, you may try using `terraform apply -replace=azapi_resource.data_collection_raw_monolog_logs_table` command to recreate instead of update. You may also refer to this [doc](https://learn.microsoft.com/en-us/azure/azure-monitor/logs/create-custom-table?tabs=azure-powershell-1%2Cazure-portal-2%2Cazure-portal-3#create-a-custom-table) for more information about custom table and DCR.

**Comment by simaotwx** [comment]
Hi @teowa , that's good to know. This sounds a lot like a workaround.

I changed some things and now I get following error:
```
│ --------------------------------------------------------------------------------
│ RESPONSE 200: 200 OK
│ ERROR CODE: SchemaUpdateFatalError
│ --------------------------------------------------------------------------------
│ {
│   "id": "/subscriptions/***/providers/Microsoft.OperationalInsights/locations/germanywestcentral/operationstatuses/***",
│   "name": "***",
│   "status": "Failed",
│   "startTime": "2023-05-24T10:10:18.7074273Z",
│   "endTime": "2023-05-24T10:10:20.1326261Z",
│   "error": {
│     "code": "SchemaUpdateFatalError",
│     "message": "HttpClient: Response status code does not indicate success - 400 (Bad Request), for request - cms/apiv2/workspaces/***/tables/RawMonologLogs_CL, due to reason - [Table validation failed on workflow creation, validation result:CMS validation failed with 1 errors.\r\nErrors:\r\n    -Workflow PopulateModelWorkflow-*** map to fact that misses mandatory fields or field type/display name not matching system configuration - Name: TimeGenerated, Display name: TimeGenerated, Type: DateTime.]."
│   }
│ }
│ --------------------------------------------------------------------------------
│ 
│ 
│   with azapi_resource.data_collection_raw_monolog_logs_table,
│   on log_analytics.tf line 26, in resource "azapi_resource" "data_collection_raw_monolog_logs_table":
│   26: resource "azapi_resource" "data_collection_raw_monolog_logs_table" {
│ 
╵
```

It seems like there must be a column called "TimeGenerated".

So I replaced "LogTime" with "TimeGenerated" and that applies.

```
azapi_resource.data_collection_raw_monolog_logs_table: Modifications complete after 22s
azurerm_monitor_data_collection_rule.log_collection_rule: Modifications complete after 3s
```

This is the new resource:


<details>
<summary>Table Resource</summary>

```terraform
resource "azapi_resource" "data_collection_raw_monolog_logs_table" {
  name      = "RawMonologLogs_CL"
  parent_id = azurerm_log_analytics_workspace.log_analytics_workspace.id
  type      = "Microsoft.OperationalInsights/workspaces/tables@2022-10-01"
  body = jsonencode(
    {
      "properties" : {
        "schema" : {
          "name" : "RawMonologLogs_CL",
          "columns" : [
            {
              "name" : "TimeGenerated",
              "type" : "datetime",
              "description" : "The time at which the data was generated"
            },
            {
              "name" : "Level",
              "type" : "string",
              "description" : "Log level"
            },
            {
              "name" : "Logger",
              "type" : "string",
              "description" : "Logger name"
            },
            {
              "name" : "Context",
              "type" : "string",
              "description" : "Context of the log line"
            },
            {
              "name" : "Message",
              "type" : "string",
              "description" : "Log message"
            },
            {
              "name" : "AdditionalContext",
              "type" : "string",
              "description" : "Additional context of the log line"
            }
          ]
        },
        "retentionInDays" : 30,
        "totalRetentionInDays" : 30
      }
    }
  )
  response_export_values = ["id"]
}
```

</details>


<details>
<summary>Rule</summary>

```terraform
resource "azurerm_monitor_data_collection_rule" "log_collection_rule" {
  name                        = "${local.deployment_name}-log-collection-dcr"
  location                    = local.location
  resource_group_name         = local.rg_name
  data_collection_endpoint_id = azurerm_monitor_data_collection_endpoint.log_collection.id

  destinations {
    log_analytics {
      workspace_resource_id = azurerm_log_analytics_workspace.log_analytics_workspace.id
      name                  = "wordpress-logs"
    }
  }

  data_flow {
    streams       = ["Custom-${azapi_resource.data_collection_raw_monolog_logs_table.name}"]
    destinations  = ["wordpress-logs"]
    output_stream = "Custom-${azapi_resource.data_collection_raw_monolog_logs_table.name}"
  }

  data_sources {
    log_file {
      name          = "wordpress-logfiles"
      format        = "text"
      streams       = ["Custom-${azapi_resource.data_collection_raw_monolog_logs_table.name}"]
      file_patterns = ["/var/local/opt/${local.deployment_name}/wordpress/volumes/logs/*.log"]
      settings {
        text {
          record_start_timestamp_format = "ISO 8601"
        }
      }
    }
  }

  stream_declaration {
    stream_name = "Custom-${azapi_resource.data_collection_raw_monolog_logs_table.name}"
    column {
      name = "TimeGenerated"
      type = "datetime"
    }
    column {
      name = "Level"
      type = "string"
    }
    column {
      name = "Logger"
      type = "string"
    }
    column {
      name = "Context"
      type = "string"
    }
    column {
      name = "Message"
      type = "string"
    }
    column {
      name = "AdditionalContext"
      type = "string"
    }
  }

  description = "Collection of logs from WordPress"
  tags        = local.default_tags
}
```

</details>

**Comment by oystesla** [comment]
I have the same issue when trying to use Microsoft-WindowsEvent. The error I'm seeing is: 
`Table for output stream 'Microsoft-WindowsEvent' is not available for destination`

**Comment by teowa** [comment]
> I have the same issue when trying to use Microsoft-WindowsEvent. The error I'm seeing is: `Table for output stream 'Microsoft-WindowsEvent' is not available for destination`
> 
> Since this isn't a custom table, I would expect this to work out of the box. I'm able to do this manually in the portal.

@oystesla , you should add below resource then the log analytic workspace can accept the `Microsoft-WindowsEvent`:
```hcl

resource "azurerm_log_analytics_solution" "example" {
  solution_name         = "WindowsEventForwarding"
  location              = azurerm_resource_group.example.location
  resource_group_name   = azurerm_resource_group.example.name
  workspace_resource_id = azurerm_log_analytics_workspace.example.id
  workspace_name        = azurerm_log_analytics_workspace.example.name
  plan {
    publisher = "Microsoft"
    product   = "OMSGallery/WindowsEventForwarding"
  }
}
```

**Comment by oystesla** [comment]
Thank you a very prompt reply, @teowa! I will give it a go. IIRC I solved this by adding an identity block to the DCR that sends data to Log Analytics, but I might have that confused. I faced the same error when I tried an event_hub destination, too - is there a similar solution for event hubs?

**Comment by KrazeyKami** [comment]
> I have the same issue when trying to use Microsoft-WindowsEvent. The error I'm seeing is: `Table for output stream 'Microsoft-WindowsEvent' is not available for destination`
> 
> Since this isn't a custom table, I would expect this to work out of the box. I'm able to do this manually in the portal.

Same problem here. I'm trying to deploy a simple DCR using ARM / JSON, with the following code (snippit):

```
"dataFlows": [
                    {
                        "streams": [
                            "Microsoft-Perf",
                            "Microsoft-Event"
                        ],
                        "destinations": [
                            "[variables('loganalyticsworkspacename')]"
                        ]
                    }
                ]
```
It returns the error ''Table for output stream 'Microsoft-WindowsEvent' is not available for destination".
If I look inside the LogAnalyticsWorkspace, I only see the default 'Perf' and 'Event' table. However, changing that in the DCR stream code gives me "Value must be one of the following values: "Microsoft-Perf", "Microsoft-InsightsMetrics"".

Clearly broken Out of Box.

**Comment by mrkesu** [comment]
> > I have the same issue when trying to use Microsoft-WindowsEvent. The error I'm seeing is: `Table for output stream 'Microsoft-WindowsEvent' is not available for destination`
>
> [...] 
>
> Same problem here. I'm trying to deploy a simple DCR using ARM / JSON, with the following code (snippit):
> 
> [...] 
> 
> It returns the error ''Table for output stream 'Microsoft-WindowsEvent' is not available for destination". If I look inside the LogAnalyticsWorkspace, I only see the default 'Perf' and 'Event' table. However, changing that in the DCR stream code gives me "Value must be one of the following values: "Microsoft-Perf", "Microsoft-InsightsMetrics"".
> 
> Clearly broken Out of Box.

This is just classic Azure Resource Management silliness. The second you want to do anything more advanced than creating a few standard resources it starts falling apart. Unfortunately everything has to go through the ARM backend eventually...

Same issue with Bicep. I just had to wait for a bit, then re-run the deployment and it works.

```bicep
param la object

targetScope = 'resourceGroup'

resource logAnalytics 'Microsoft.OperationalInsights/workspaces@2023-09-01' = {
  location: la.location
  name: la.workspaceName
  properties: {
    sku: {
      name: la.serviceTier
    }
    retentionInDays: la.dataRetention
  }
}


// This process involves another "long running operation" which means this part will fail on first run.
// If it fails, just wait a few minutes and run again... 😒
// to-do: split this up in GitHub Actions so we can put in some kind of verification between executions or an artificial wait timer like usual...
resource dcr 'Microsoft.Insights/dataCollectionRules@2023-03-11' = {
  name: la.workspaceDCRName
  location: la.location
  kind: 'Windows'
  properties: {
    dataSources: {
      performanceCounters: [
        {
          streams: [
            'Microsoft-Perf'
          ]
          samplingFrequencyInSeconds: 30
          counterSpecifiers: [
            '\\LogicalDisk(C:)\\Avg. Disk Queue Length'
            '\\LogicalDisk(C:)\\Current Disk Queue Length'
            '\\Memory\\Available Mbytes'
            '\\Memory\\Page Faults/sec'
            '\\Memory\\Pages/sec'
            '\\Memory\\% Committed Bytes In Use'
            '\\PhysicalDisk(*)\\Avg. Disk Queue Length'
            '\\PhysicalDisk(*)\\Avg. Disk sec/Read'
            '\\PhysicalDisk(*)\\Avg. Disk sec/Transfer'
            '\\PhysicalDisk(*)\\Avg. Disk sec/Write'
            '\\Processor Information(_Total)\\% Processor Time'
            '\\User Input Delay per Process(*)\\Max Input Delay'
            '\\User Input Delay per Session(*)\\Max Input Delay'
            '\\RemoteFX Network(*)\\Current TCP RTT'
            '\\RemoteFX Network(*)\\Current UDP Bandwidth'
          ]
          name: 'perfCounterDataSource30'
        }
        {
          streams: [
            'Microsoft-Perf'
          ]
          samplingFrequencyInSeconds: 60
          counterSpecifiers: [
            '\\LogicalDisk(C:)\\% Free Space'
            '\\LogicalDisk(C:)\\Avg. Disk sec/Transfer'
          ]
          name: 'perfCounterDataSource60'
        }
      ]
      windowsEventLogs: [
        {
          streams: [
            'Microsoft-Event'
          ]
          xPathQueries: [
            'Microsoft-Windows-TerminalServices-RemoteConnectionManager/Admin!*[System[(Level=2 or Level=3 or Level=4 or Level=0)]]'
            'Microsoft-Windows-TerminalServices-LocalSessionManager/Operational!*[System[(Level=2 or Level=3 or Level=4 or Level=0)]]'
            'System!*'
            'Microsoft-FSLogix-Apps/Operational!*[System[(Level=2 or Level=3 or Level=4 or Level=0)]]'
            'Application!*[System[(Level=2 or Level=3)]]'
            'Microsoft-FSLogix-Apps/Admin!*[System[(Level=2 or Level=3 or Level=4 or Level=0)]]'
          ]
          name: 'eventLogsDataSource'
        }
      ]
    }
    destinations: {
      logAnalytics: [
        {
          workspaceResourceId: logAnalytics.id
          name: logAnalytics.properties.
        }
      ]
    }
    dataFlows: [
      {
        streams: [
          'Microsoft-Perf'
          'Microsoft-Event'
        ]
        destinations: [
          logAnalytics.name
        ]
      }
    ]
  }
}
```
First execution results in errors:

```
 - Data collection rule is invalid (Code: InvalidPayload)
   - Table for output stream 'Microsoft-Perf' is not available for destination 'avd-dev-log-v2-ws'. (Code:InvalidOutputTable)
   - Table for output stream 'Microsoft-Event' is not available for destination 'avd-dev-log-v2-ws'. (Code:InvalidOutputTable)
```

Wait for a few minutes, then retry the deployment and it's successful.

_I would create an issue in MS's GitHub repo, but from experience I've learned that it does nothing (still waiting on several simple issues reported **_years_** ago)_

**Comment by rich-debug** [comment]
After following the solution above to create a custom logs table im getting the below error 
` Error: checking for presence of existing Resource: (ResourceId "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxx/xxxxxxx/xxxxxxxxx/xxxxxxxxx/xxxxxxxxxxxx/xxxxxxxxxxxxxxxxxxxx/xxxxxx/xxxxxxxx-logs-table" / Api Version "2022-10-01"): ChainedTokenCredential authentication failed
│ GET http://xxx.xxx.xxx.xxx/metadata/identity/oauth2/token
│ --------------------------------------------------------------------------------
│ RESPONSE 401 Unauthorized
│ --------------------------------------------------------------------------------
│ Response contained no body
│ --------------------------------------------------------------------------------
│ 
│ 
│   with azapi_resource.data_collection_logs_table,
│   on logs-table.tf line 8, in resource "azapi_resource" "data_collection_logs_table":
│    8: resource "azapi_resource" "data_collection_logs_table" {`

**Comment by KrazeyKami** [comment]
> > > I have the same issue when trying to use Microsoft-WindowsEvent. The error I'm seeing is: `Table for output stream 'Microsoft-WindowsEvent' is not available for destination`
> > 
> > 
> > [...]
> > Same problem here. I'm trying to deploy a simple DCR using ARM / JSON, with the following code (snippit):
> > [...]
> > It returns the error ''Table for output stream 'Microsoft-WindowsEvent' is not available for destination". If I look inside the LogAnalyticsWorkspace, I only see the default 'Perf' and 'Event' table. However, changing that in the DCR stream code gives me "Value must be one of the following values: "Microsoft-Perf", "Microsoft-InsightsMetrics"".
> > Clearly broken Out of Box.
> 
> This is just classic Azure Resource Management silliness. The second you want to do anything more advanced than creating a few standard resources it starts falling apart. Unfortunately everything has to go through the ARM backend eventually...
> 
> Same issue with Bicep. I just had to wait for a bit, then re-run the deployment and it works.
> 
> ```bicep
> param la object
> 
> targetScope = 'resourceGroup'
> 
> resource logAnalytics 'Microsoft.OperationalInsights/workspaces@2023-09-01' = {
>   location: la.location
>   name: la.workspaceName
>   properties: {
>     sku: {
>       name: la.serviceTier
>     }
>     retentionInDays: la.dataRetention
>   }
> }
> 
> 
> // This process involves another "long running operation" which means this part will fail on first run.
> // If it fails, just wait a few minutes and run again... 😒
> // to-do: split this up in GitHub Actions so we can put in some kind of verification between executions or an artificial wait timer like usual...
> resource dcr 'Microsoft.Insights/dataCollectionRules@2023-03-11' = {
>   name: la.workspaceDCRName
>   location: la.location
>   kind: 'Windows'
>   properties: {
>     dataSources: {
>       performanceCounters: [
>         {
>           streams: [
>             'Microsoft-Perf'
>           ]
>           samplingFrequencyInSeconds: 30
>           counterSpecifiers: [
>             '\\LogicalDisk(C:)\\Avg. Disk Queue Length'
>             '\\LogicalDisk(C:)\\Current Disk Queue Length'
>             '\\Memory\\Available Mbytes'
>             '\\Memory\\Page Faults/sec'
>             '\\Memory\\Pages/sec'
>             '\\Memory\\% Committed Bytes In Use'
>             '\\PhysicalDisk(*)\\Avg. Disk Queue Length'
>             '\\PhysicalDisk(*)\\Avg. Disk sec/Read'
>             '\\PhysicalDisk(*)\\Avg. Disk sec/Transfer'
>             '\\PhysicalDisk(*)\\Avg. Disk sec/Write'
>             '\\Processor Information(_Total)\\% Processor Time'
>             '\\User Input Delay per Process(*)\\Max Input Delay'
>             '\\User Input Delay per Session(*)\\Max Input Delay'
>             '\\RemoteFX Network(*)\\Current TCP RTT'
>             '\\RemoteFX Network(*)\\Current UDP Bandwidth'
>           ]
>           name: 'perfCounterDataSource30'
>         }
>         {
>           streams: [
>             'Microsoft-Perf'
>           ]
>           samplingFrequencyInSeconds: 60
>           counterSpecifiers: [
>             '\\LogicalDisk(C:)\\% Free Space'
>             '\\LogicalDisk(C:)\\Avg. Disk sec/Transfer'
>           ]
>           name: 'perfCounterDataSource60'
>         }
>       ]
>       windowsEventLogs: [
>         {
>           streams: [
>             'Microsoft-Event'
>           ]
>           xPathQueries: [
>             'Microsoft-Windows-TerminalServices-RemoteConnectionManager/Admin!*[System[(Level=2 or Level=3 or Level=4 or Level=0)]]'
>             'Microsoft-Windows-TerminalServices-LocalSessionManager/Operational!*[System[(Level=2 or Level=3 or Level=4 or Level=0)]]'
>             'System!*'
>             'Microsoft-FSLogix-Apps/Operational!*[System[(Level=2 or Level=3 or Level=4 or Level=0)]]'
>             'Application!*[System[(Level=2 or Level=3)]]'
>             'Microsoft-FSLogix-Apps/Admin!*[System[(Level=2 or Level=3 or Level=4 or Level=0)]]'
>           ]
>           name: 'eventLogsDataSource'
>         }
>       ]
>     }
>     destinations: {
>       logAnalytics: [
>         {
>           workspaceResourceId: logAnalytics.id
>           name: logAnalytics.name
>         }
>       ]
>     }
>     dataFlows: [
>       {
>         streams: [
>           'Microsoft-Perf'
>           'Microsoft-Event'
>         ]
>         destinations: [
>           logAnalytics.name
>         ]
>       }
>     ]
>   }
> }
> ```
> 
> First execution results in errors:
> 
> ```
>  - Data collection rule is invalid (Code: InvalidPayload)
>    - Table for output stream 'Microsoft-Perf' is not available for destination 'avd-dev-log-v2-ws'. (Code:InvalidOutputTable)
>    - Table for output stream 'Microsoft-Event' is not available for destination 'avd-dev-log-v2-ws'. (Code:InvalidOutputTable)
> ```
> 
> Wait for a few minutes, then retry the deployment and it's successful.
> 
> _I would create an issue in MS's GitHub repo, but from experience I've learned that it does nothing (still waiting on several simple issues reported **_years_** ago)_

I can confirm also with Bicep, as of today (1st November 2024), this is still an issue / broken.
Did several deployments without issues, and suddenly it starts failing:

```
New-AzDeployment: 15:14:08 - The deployment 'AVD' failed with error(s). Showing 1 out of 1 error(s).
Status Message: At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details. (Code: DeploymentFailed)
 - Data collection rule is invalid (Code: InvalidPayload)
   - Table for output stream 'Microsoft-Perf' is not available for destination 'AVD-LAW'. (Code:InvalidOutputTable)
   - Table for output stream 'Microsoft-Event' is not available for destination 'AVD-LAW'. (Code:InvalidOutputTable)
```
